### ISSUE TITLE
Array.prototype.map() allows assigning additional properties to an array of defined type

### ISSUE THREAD

**Issue by esanzgar** [microsoft/TypeScript#36354]
**TypeScript Versions:**  from 3.5.1 to Nightly (issue not present in 3.3.3)

**Search Terms:** `array map`

**Expected behavior:**  Third assignment should raise an error

**Actual behavior:** Assignment is allowed

**Related Issues:** No found any

**Code**
```ts
interface A {
    a: 2
}

// 1. Correct: b property is not allowed in type A
const a: A = { a: 2, b: 8 }

// 2. Correct: b property is not allowed in type A
const other: A[] = [{ a: 2, b: 8 }]

// 3. Incorrect (no error): why is b property permitted here?
const array: A[] = [1, 2].map(i => ({ a: 2, b: 8 }))
```

<details><summary><b>Output</b></summary>

```ts
"use strict";
// 1. Correct: b property is not allowed in type A
const a = { a: 2, b: 8 };
// 2. Correct: b property is not allowed in type A
const other = [{ a: 2, b: 8 }];
// 3. Incorrect (no error): why is b property permitted here?
const array = [1, 2].map(i => ({ a: 2, b: 8 }));

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "useDefineForClassFields": false,
    "alwaysStrict": true,
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "downlevelIteration": false,
    "noEmitHelpers": false,
    "noLib": false,
    "noStrictGenericChecks": false,
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "esModuleInterop": true,
    "preserveConstEnums": false,
    "removeComments": false,
    "skipLibCheck": false,
    "checkJs": false,
    "allowJs": false,
    "declaration": true,
    "experimentalDecorators": false,
    "emitDecoratorMetadata": false,
    "target": "ES2017",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](http://www.typescriptlang.org/play/?ts=3.8.0-dev.20200119#)

**Comment by jcalz** [comment]
Duplicate of #33158


The error in 3.3.3 is not the one you're looking for.  That error is due to widening of `2` to `number`; it has nothing to do with the presence of the `b` property.  Even this errors in 3.3.3:
```ts
const array: A[] = [1, 2].map(i => ({ a: 2 })); // error!
// Type '{ a: number; }' is not assignable to type 'A'.
```

In reality the error you expect is not happening because [excess property checking](https://www.typescriptlang.org/docs/handbook/interfaces.html#excess-property-checks) only happens in very specific circumstances.  See #33158 for more discussion.  Good luck!

**Comment by IllusionMH** [comment]
Or duplicate of #241
